### PR TITLE
Jerry API for array creation and manipulation

### PR DESCRIPTION
**Pull request by chunseoklee**
1. jerry_api_create_array_object : create array object of given size
2. jerry_api_set_array_index_value : set value at given index
3. jerry_api_get_array_index_value : get value at given index

JerryScript-DCO-1.0-Signed-off-by: chunseok lee chunseok.lee@samsung.com

### CODE REVIEW

**Comment by chunseoklee** [comment]
@egavrin Aside from this commit, do we need to discuss each jerry API before implementing it  ?

**Comment by egavrin** [comment]
@chunseoklee only, if there're any questions or change in already existing APIs.

File: jerry-core/jerry.cpp [hunk 1]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+ *
+ * Note:
+ *      if value was retrieved successfully, it should be freed
+ *      with jerry_api_release_value just when it becomes unnecessary.
+ *
+ * @return true, if field value was retrieved successfully, i.e. upon the call:
+ *                - there is field with specified name in the object;
+ *         false - otherwise.
+ */
+bool jerry_api_get_array_index_value (jerry_api_object_t *array_obj_p,
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: This function looks like the previous one. Maybe we can merge them?

File: jerry-core/jerry.cpp [hunk 2]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
 
 /**
+ * Create an array object
+ *
+ * Note:
+ *      caller should release the object with jerry_api_release_object, just when the value becomes unnecessary.
+ *
+ * @return pointer to created array object
+ */
+jerry_api_object_t*
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: `jerry_api_object_t *`

File: jerry-core/jerry.cpp [hunk 3]
```diff
@@ -24,6 +24,7 @@
 #include "ecma-helpers.h"
 #include "ecma-init-finalize.h"
 #include "ecma-objects.h"
+#include "ecma-array-object.h"
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Headers should be in alphabetical order.

File: tests/unit/test-api.cpp [hunk 4]
```diff
@@ -516,6 +516,23 @@ main (void)
...
+  // Test: Array Object API
+  jerry_api_object_t *array_objp = jerry_api_create_array_object (10);
+
+  jerry_api_value_t v_in;
+  test_api_init_api_value_float64 (&v_in, 10.5);
+  jerry_api_set_array_index_value (array_objp, 5, &v_in);
+  jerry_api_value_t v_out;
+  jerry_api_get_array_index_value (array_objp, 5, &v_out);
+
+  JERRY_ASSERT (v_out.type==JERRY_API_DATA_TYPE_FLOAT64 && v_out.v_float64 == 10.5);
```
Comment (egavrin) [inline tests/unit/test-api.cpp]: `v_out.type == JERRY_API_...`

File: tests/unit/test-api.cpp [hunk 5]
```diff
@@ -516,6 +516,23 @@ main (void)
 
   jerry_api_release_object (obj_p);
 
+
+  // Test: Array Object API
+  jerry_api_object_t *array_objp = jerry_api_create_array_object (10);
```
Comment (egavrin) [inline tests/unit/test-api.cpp]: `array_obj_p`

File: jerry-core/jerry.cpp [hunk 6]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+bool jerry_api_get_array_index_value (jerry_api_object_t *array_obj_p,
+                                      jerry_api_length_t index,
+                                      jerry_api_value_t *value_p) /* output to put value */
+{
+  ecma_number_t index_n = ecma_uint32_to_number (index);
+  lit_utf8_byte_t str[64];
+  ecma_number_to_utf8_string (index_n, str, sizeof (str));
+
+  return jerry_api_get_object_field_value (array_obj_p, str, value_p);
+
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Extra empty line.

File: jerry-core/jerry.cpp [hunk 7]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+  ecma_property_t *length_prop_p = ecma_create_named_data_property (obj_p,
+                                                                    length_magic_string_p,
+                                                                    true, false, false);
+  ecma_set_named_data_property_value (length_prop_p, ecma_make_number_value (length_num_p));
+
+  ecma_deref_ecma_string (length_magic_string_p);
+
+  return obj_p;
+
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Extra empty line.

File: jerry-core/jerry.cpp [hunk 8]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+} /* jerry_api_create_array */
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Incorrect comment.

File: jerry-core/jerry.cpp [hunk 9]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+bool jerry_api_set_array_index_value (jerry_api_object_t *array_obj_p,
+                                      jerry_api_length_t index,
+                                      jerry_api_value_t *value_p)
+{
+  ecma_number_t index_n = ecma_uint32_to_number (index);
+  lit_utf8_byte_t str[64];
+  ecma_number_to_utf8_string (index_n, str, sizeof (str));
+
+  return jerry_api_set_object_field_value (array_obj_p, str, value_p);
+} /* jerry_api_set_array_value */
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Incorrect comment

File: jerry-core/jerry.cpp [hunk 10]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+                                      jerry_api_length_t index,
+                                      jerry_api_value_t *value_p) /* output to put value */
+{
+  ecma_number_t index_n = ecma_uint32_to_number (index);
+  lit_utf8_byte_t str[64];
+  ecma_number_to_utf8_string (index_n, str, sizeof (str));
+
+  return jerry_api_get_object_field_value (array_obj_p, str, value_p);
+
+} /* jerry_api_get_array_value */
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Incorrect comment

File: jerry-core/jerry.cpp [hunk 11]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
 /**
+ * Create an array object
+ *
+ * Note:
+ *      caller should release the object with jerry_api_release_object, just when the value becomes unnecessary.
+ *
+ * @return pointer to created array object
+ */
+jerry_api_object_t*
+jerry_api_create_array_object (jerry_api_size_t size)
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Please add comments for parameters.

File: jerry-core/jerry.cpp [hunk 12]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+} /* jerry_api_create_array */
+
+
+/**
+ * Set value of field in the specified array object
+ *
+ * @return true, if field value was set successfully
+ *         false - otherwise.
+ */
+bool jerry_api_set_array_index_value (jerry_api_object_t *array_obj_p,
```
Comment (egavrin) [inline jerry-core/jerry.cpp]: Please add comments for parameters.

File: jerry-core/jerry.cpp (hunk 1 above)
Comment (egavrin) [inline jerry-core/jerry.cpp]: Please add comments for parameters.

Comment (egavrin) [inline jerry-core/jerry.cpp]:
```c
bool
jerry_api_get_array_index_value (jerry_api_object_t *array_obj_p,
```

File: jerry-core/jerry.cpp (hunk 12 above)
Comment (egavrin) [inline jerry-core/jerry.cpp]:
```c
bool
jerry_api_set_array_index_value (jerry_api_object_t *array_obj_p,
```

File: jerry-core/jerry.cpp [hunk 13]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+{
+  jerry_api_size_t length = size;
+
+#ifndef CONFIG_ECMA_COMPACT_PROFILE_DISABLE_ARRAY_BUILTIN
+  ecma_object_t *array_prototype_obj_p = ecma_builtin_get (ECMA_BUILTIN_ID_ARRAY_PROTOTYPE);
+#else /* !CONFIG_ECMA_COMPACT_PROFILE_DISABLE_ARRAY_BUILTIN */
+  ecma_object_t *array_prototype_obj_p = ecma_builtin_get (ECMA_BUILTIN_ID_OBJECT_PROTOTYPE);
+#endif /* CONFIG_ECMA_COMPACT_PROFILE_DISABLE_ARRAY_BUILTIN */
+
+  ecma_object_t *obj_p = ecma_create_object (array_prototype_obj_p, true, ECMA_OBJECT_TYPE_ARRAY);
```
Comment (ruben-ayrapetyan) [inline jerry-core/jerry.cpp]: We could use `ecma_op_create_array_object` for creating array in the case.

File: jerry-core/jerry.cpp [hunk 14]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+ */
+bool jerry_api_set_array_index_value (jerry_api_object_t *array_obj_p,
+                                      jerry_api_length_t index,
+                                      jerry_api_value_t *value_p)
+{
+  ecma_number_t index_n = ecma_uint32_to_number (index);
+  lit_utf8_byte_t str[64];
+  ecma_number_to_utf8_string (index_n, str, sizeof (str));
+
+  return jerry_api_set_object_field_value (array_obj_p, str, value_p);
```
Comment (ruben-ayrapetyan) [inline jerry-core/jerry.cpp]: In the case, integer index could be passed without converting it through character buffer, like in the https://github.com/Samsung/jerryscript/commit/cdf361c2bd2fd6d0a014c665e82b820f0c96cc85#diff-aa76478615f09c89ff65ab86bf8ebe53R1718

File: jerry-core/jerry.cpp [hunk 15]
```diff
@@ -430,6 +431,85 @@ jerry_api_create_object (void)
...
+ */
+bool jerry_api_get_array_index_value (jerry_api_object_t *array_obj_p,
+                                      jerry_api_length_t index,
+                                      jerry_api_value_t *value_p) /* output to put value */
+{
+  ecma_number_t index_n = ecma_uint32_to_number (index);
+  lit_utf8_byte_t str[64];
+  ecma_number_to_utf8_string (index_n, str, sizeof (str));
+
+  return jerry_api_get_object_field_value (array_obj_p, str, value_p);
```
Comment (ruben-ayrapetyan) [inline jerry-core/jerry.cpp]: The conversion through buffer could also be removed, using `ecma_new_ecma_string_from_uint32` and `ecma_op_object_get`.

**Comment by chunseoklee** [comment]
@egavrin @ruben-ayrapetyan Updated as commented.

**Comment by egavrin** [comment]
Good to me

File: tests/unit/test-api.cpp [hunk 16]
```diff
@@ -516,6 +516,23 @@ main (void)
...
+  jerry_api_value_t v_out;
+  jerry_api_get_array_index_value (array_obj_p, 5, &v_out);
+
+  JERRY_ASSERT (v_out.type == JERRY_API_DATA_TYPE_FLOAT64 && v_out.v_float64 == 10.5);
+
+  jerry_api_release_object (array_obj_p);
+
+
+
+
```
Comment (zherczeg) [inline tests/unit/test-api.cpp]: Too many empty lines?

File: jerry-core/jerry.cpp [hunk 17]
```diff
@@ -430,6 +431,106 @@ jerry_api_create_object (void)
...
+    is_successful = false;
+  }
+
+  ecma_free_completion_value (get_completion);
+  ecma_deref_ecma_string (str_idx_p);
+
+  return is_successful;
+} /* jerry_api_get_array_index_value */
+
+
```
Comment (zherczeg) [inline jerry-core/jerry.cpp]: Only one newline

File: jerry-core/jerry.cpp [hunk 18]
```diff
@@ -430,6 +431,106 @@ jerry_api_create_object (void)
...
+  if (!ecma_is_completion_value_normal (get_completion))
+  {
+    JERRY_ASSERT (ecma_is_completion_value_throw (get_completion));
+  }
+
+  ecma_free_value (array_length, true);
+  return obj_p;
+} /* jerry_api_create_array_object */
+
+
```
Comment (zherczeg) [inline jerry-core/jerry.cpp]: Only one newline

**Comment by zherczeg** [comment]
LGTM

**Comment by chunseoklee** [comment]
@zherczeg Updated as commented. Please check it.

File: jerry-core/jerry.cpp [hunk 19]
```diff
@@ -430,6 +431,104 @@ jerry_api_create_object (void)
...
+  ecma_value_t array_length = ecma_make_number_value (length_num_p);
+
+  jerry_api_length_t argument_size = 1;
+  ecma_completion_value_t get_completion = ecma_op_create_array_object (&array_length, argument_size, true);
+  ecma_value_t val = ecma_get_completion_value_value (get_completion);
+  jerry_api_object_t *obj_p = ecma_get_object_from_value (val);
+
+  if (!ecma_is_completion_value_normal (get_completion))
+  {
+    JERRY_ASSERT (ecma_is_completion_value_throw (get_completion));
```
Comment (ruben-ayrapetyan) [inline jerry-core/jerry.cpp]: Exception couldn't be thrown by `ecma_op_create_array_object`, so this `if` block could be replaced with assertion like `JERRY_ASSERT (ecma_is_completion_value_normal (new_array_completion));` before line 452.

File: jerry-core/jerry.cpp [hunk 20]
```diff
@@ -430,6 +431,104 @@ jerry_api_create_object (void)
...
+{
+  bool is_successful = true;
+  ecma_string_t *str_idx_p = ecma_new_ecma_string_from_uint32 ((uint32_t) index);
+  ecma_value_t value;
+  jerry_api_convert_api_value_to_ecma_value (&value, value_p);
+  ecma_completion_value_t get_completion = ecma_op_object_put (array_obj_p, str_idx_p, value, false);
+
+  if (!ecma_is_completion_value_normal (get_completion))
+  {
+    JERRY_ASSERT (ecma_is_completion_value_throw (get_completion));
```
Comment (ruben-ayrapetyan) [inline jerry-core/jerry.cpp]: If `is_throw` argument of `ecma_op_object_put` is `false`, exception can't be thrown by the function.
In the case, the `if` block could also be replaced with assertion that completion value is normal.

**Comment by chunseoklee** [comment]
@ruben-ayrapetyan Updated as you commented.

**Comment by ruben-ayrapetyan** [comment]
@chunseoklee, thank you for the update.

**Comment by ruben-ayrapetyan** [comment]
Looks good to me

**Comment by egavrin** [comment]
`make push`

**Pull request merged**